rmw_opendds_cpp/src/DDSSubscriber.cpp

```diff
@@ -0,0 +1,179 @@
+// Copyright 2014-2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rmw_opendds_cpp/DDSSubscriber.hpp"
+#include "opendds_static_serialized_dataTypeSupportImpl.h"
+
+#include "rmw_opendds_shared_cpp/event_converter.hpp"
+#include "rmw_opendds_shared_cpp/identifier.hpp"
+#include "rmw_opendds_shared_cpp/qos.hpp"
+
+#include <dds/DCPS/DataReaderImpl_T.h>
+
+#include "rmw/visibility_control.h"
+#include "rmw/incompatible_qos_events_statuses.h"
+
+DDSSubscriber * DDSSubscriber::from(const rmw_subscription_t * sub)
+{
+  if (!sub) {
+    RMW_SET_ERROR_MSG("rmw_subscription is null");
+    return nullptr;
+  }
+  if (!check_impl_id(sub->implementation_identifier)) {
+    return nullptr;
+  }
+  auto dds_sub = static_cast<DDSSubscriber *>(sub->data);
+  if (!dds_sub) {
+    RMW_SET_ERROR_MSG("dds_sub is null");
+    return nullptr;
+  }
+  return dds_sub;
+}
+
+rmw_ret_t DDSSubscriber::get_rmw_qos(rmw_qos_profile_t & qos) const
+{
+  DDS::DataReaderQos dds_qos;
+  if (reader_->get_qos(dds_qos) == DDS::RETCODE_OK) {
+    dds_qos_to_rmw_qos(dds_qos, qos);
+    return RMW_RET_OK;
+  } else {
+    RMW_SET_ERROR_MSG("get_rmw_qos failed");
+  }
+  return RMW_RET_ERROR;
+}
+
+rmw_ret_t DDSSubscriber::to_ros_message(const rcutils_uint8_array_t & cdr_stream, void * ros_message)
+{
+  if (callbacks_->to_message(&cdr_stream, ros_message)) {
+    return RMW_RET_OK;
+  }
+  RMW_SET_ERROR_MSG("to_ros_message failed");
+  return RMW_RET_ERROR;
+}
+
+rmw_ret_t DDSSubscriber::get_status(const DDS::StatusMask mask, void * rmw_status)
+{
+  switch (mask) {
+    case DDS::LIVELINESS_CHANGED_STATUS:
+    {
+      DDS::LivelinessChangedStatus llc;
+      rmw_ret_t ret = check_dds_ret_code(reader_->get_liveliness_changed_status(llc));
+      if (ret != RMW_RET_OK) {
+        return ret;
+      }
+      auto rmw_llc = static_cast<rmw_liveliness_changed_status_t *>(rmw_status);
+      rmw_llc->alive_count = llc.alive_count;
```

Comment:
Yes, they are different between pub and sub. For example, 
one cares about OFFERED_DEADLINE_MISSED_STATUS, 
while the other REQUESTED_DEADLINE_MISSED_STATUS.

However, mistakes/typos may occur in situations like this. I tried to avoid them, but you still caught one in another place. I appreciate it. So please do the same on these two functions. Thanks.